url5a01f2c1d9_core9_core_tgz-pJ8-io_core9_core_boot_BootstrapFrameworkJ8/src/net/xeoh/plugins/base/impl/classpath/cache/JARCache.java

```diff
@@ -0,0 +1,298 @@
+		public boolean classesValid = false;
+
+	}
+
+	/** */
+	final Logger logger = Logger.getLogger(this.getClass().getName());
+
+	/** Name of the default cache file */
+	private final static String DEFAULT_CACHE_FILE = "jspf.plugin.cache";
+
+	/** Is the cache enabled */
+	private boolean cacheEnabled = false;
+
+	/** maps a fingerprint to a jar information */
+	private Map<String, JARInformation> cacheMap = new HashMap<String, JARInformation>();
+
+	/** */
+	private boolean weakMode = false;
+
+	private String cachePath;
+
+	/**
+	 * Load cache
+	 */
+	@SuppressWarnings("unchecked")
+	public void loadCache() {
+		if (!this.cacheEnabled)
+			return;
+
+		final String cacheFile = (this.cachePath == null) ? DEFAULT_CACHE_FILE : this.cachePath;
+
+		try {
+			// lock the lockfile
+			final FileInputStream fis = new FileInputStream(cacheFile);
+			final ObjectInputStream ois = new ObjectInputStream(fis);
+			final Object rval = ois.readObject(); // Loading caches can take
+												  // very long
+			                                      // (>250ms) for many entries
+
+			ois.close();
+
+			if (rval != null) {
+				this.cacheMap = (Map<String, JARInformation>) rval;
+			}
+		} catch (EOFException e) {
+			//
+		} catch (final FileNotFoundException e) {
+			// If file was not found, no problem ...
+			// e.printStackTrace();
+		} catch (final IOException e) {
+			e.printStackTrace();
+		} catch (final ClassNotFoundException e) {
+			this.logger
+			        .warning("Your JSPF cache is outdated, please delete it. It will be regenerated with the next run. The next exception reflects this, so don't be afraid.");
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * @param uri
+	 * @return .
+	 */
+	public JARInformation getJARInformationFor(final URI uri) {
+		// Just return a dummy when we are disabled
+		if (!this.cacheEnabled) {
+			return new JARInformation();
+		}
+
+		final File file = new File(uri);
+		final String hash = (this.weakMode) ? generateWeakHash(file) : generateStrongHash(file);
+
+		if (hash == null) {
+			this.logger.warning("Error generating hash. Caching won't work.");
+			return new JARInformation();
+		}
+
+		// Create information if not already there.
+		if (!this.cacheMap.containsKey(hash)) {
+			this.cacheMap.put(hash, new JARInformation());
+		}
+
+		final JARInformation jarInformation = this.cacheMap.get(hash);
+		jarInformation.lastAccess = System.currentTimeMillis();
+		jarInformation.usageCount++;
+		return jarInformation;
+	}
+
+	/**
+	 * Saves the cache
+	 */
+	public void saveCache() {
+
+		if (!this.cacheEnabled)
+			return;
+		String cacheFile = this.cachePath;
+		// String = this.configuration.getConfiguration(getClass(),
+		// "cacheFile");
+		if (cacheFile == null) {
+			cacheFile = DEFAULT_CACHE_FILE;
+		}
+
+		try {
+			// lock the lockfile
+			final FileOutputStream fos = new FileOutputStream(cacheFile);
+			@SuppressWarnings("resource")
+			final ObjectOutputStream oos = new ObjectOutputStream(fos);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *oos*. Execution paths that do not contain closure statements were detected. Ensure *oos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).